### PR TITLE
feat: Add batch_create_links method for creating cross-references between objects

### LINKED ISSUE #42568: Add batch_create_links method in Weaviate hook
### Description

Add the `batch_create_links` method in the Weaviate hook to allow users batch create links from an object to anothe object through cross-references. 

An example will be:

```
data = {
                    "collection_name": "TestCollection1",
                    "from_property": "property1",
                    "from_uuid": 1234455,
                    "to_uuid": 1245555,
                }
weaviate_hook.batch_create_links(data)
```

### Use case/motivation

Currently basic batch import is already implemented. Using this method, users can create links between objects and also assign `tenant` as a kwarg.

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by harjeevanmaan**
- handles both `list `and `DataFrame `data
- Added proper validation and error messages for `collection_name`, `from_property`, `from_uuid`, and `to_uuid `fields.
- Ensure fields are present and of the correct type before attempting to add cross-references.
- Retry logic with Retrying mechanism

test: Unit test for batch_create_links method

- Covered scenarios for both `list `and `DataFrame `data input.
- Validated the correct number of reference creations using mock object.

Fixes: #42568

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by harjeevanmaan** [comment]
@Lee-W @utkarsharma2 Any updates on this?

**Comment by Lee-W** [comment]
> @Lee-W @utkarsharma2 Any updates on this?

Will take a deeper look today

**Comment by harjeevanmaan** [comment]
Thanks!

File: airflow/providers/weaviate/hooks/weaviate.py [hunk 1]
```diff
@@ -315,6 +315,88 @@ def batch_data(
                         )
                         self.log.debug("Inserted object with uuid: %s into batch", uuid)
 
+    def batch_create_links(
+        self,
+        data: list[dict[str, Any]] | pd.DataFrame,
+        collection_name: str = "",
+        batch_size: int = 100,
+        concurrent_requests: int = 10,
+        consistency_level: Any = None,
```
Comment (Lee-W) [inline airflow/providers/weaviate/hooks/weaviate.py:324]: What is this? Would it be possible for us to narrow the type here?

File: airflow/providers/weaviate/hooks/weaviate.py [hunk 2]
```diff
@@ -315,6 +315,88 @@ def batch_data(
...
+
+        for data_obj in converted_data:
+            if collection_name == "":
+                collection_name = data_obj.pop("collection_name", None)
+                if collection_name is None:
+                    raise ValueError("collection_name is a required field")
+                elif not isinstance(collection_name, str) or collection_name == "":
+                    raise ValueError("collection_name should be a valid string")
+                else:
+                    pass
```
Comment (Lee-W) [inline airflow/providers/weaviate/hooks/weaviate.py:350]:
```suggestion
```

We probably don't need this

File: airflow/providers/weaviate/hooks/weaviate.py [hunk 3]
```diff
@@ -315,6 +315,88 @@ def batch_data(
...
+        collection_name_given = collection_name
+        converted_data = self._convert_dataframe_to_list(data)
+
+        for data_obj in converted_data:
+            if collection_name == "":
+                collection_name = data_obj.pop("collection_name", None)
+                if collection_name is None:
+                    raise ValueError("collection_name is a required field")
+                elif not isinstance(collection_name, str) or collection_name == "":
+                    raise ValueError("collection_name should be a valid string")
```
Comment (Lee-W) [inline airflow/providers/weaviate/hooks/weaviate.py:348]: technically "" is a valid string 🤔 a non-empty string maybe?
```suggestion
                    raise ValueError("collection_name should be a non-empty string")
```

File: airflow/providers/weaviate/hooks/weaviate.py [hunk 4]
```diff
@@ -315,6 +315,88 @@ def batch_data(
...
+                    stop=stop_after_attempt(retry_attempts_per_object),
+                    retry=(
+                        retry_if_exception(lambda exc: check_http_error_is_retryable(exc))
+                        | retry_if_exception_type(REQUESTS_EXCEPTIONS_TYPES)
+                    ),
+                ):
+                    with attempt:
+                        try:
+                            from_property = data_obj.pop("from_property", None)
+                            from_uuid = data_obj.pop("from_uuid")
```
Comment (Lee-W) [inline airflow/providers/weaviate/hooks/weaviate.py:367]: do we need to fallback to None here?

File: airflow/providers/weaviate/hooks/weaviate.py [hunk 5]
```diff
@@ -315,6 +315,88 @@ def batch_data(
...
+                                    from_uuid=from_uuid,
+                                    to=to_uuid,
+                                    tenant=tenant,
+                                )
+                                self.log.debug(
+                                    "Created links between object with uuid: %s and object with uuid: %s",
+                                    from_uuid,
+                                    to_uuid,
+                                )
+                                collection_name = collection_name_given
```
Comment (Lee-W) [inline airflow/providers/weaviate/hooks/weaviate.py:395]:
```suggestion
                            self.log.debug(
                                "Attempt %s of referencing an object to another object: %s",
                                attempt.retry_state.attempt_number,
                                from_uuid,
                            )
                            batch.add_reference(
                                from_collection=collection_name,
                                from_property=from_property,
                                from_uuid=from_uuid,
                                to=to_uuid,
                                tenant=tenant,
                            )
                            self.log.debug(
                                "Created links between object with uuid: %s and object with uuid: %s",
                                from_uuid,
                                to_uuid,
                            )
                            collection_name = collection_name_given
```

File: airflow/providers/weaviate/hooks/weaviate.py [hunk 6]
```diff
@@ -315,6 +315,88 @@ def batch_data(
...
+                        except Exception as e:
```
Comment (Lee-W) [inline airflow/providers/weaviate/hooks/weaviate.py:396]: let's catch a narrower excetipn

**Pull request closed without merging**